### DISCUSSION TITLE
aggioranmento python3

### DISCUSSION THREAD

**Discussion by sp2003** [indomus/forum#1727] [HUB: Home Assistant]
ciao a tutti

dovevo aggiornare una installazione di HA Core e passare a una versione superiore di python3.8.
coem d guida ho aggiornato:

```
sudo apt install rustc
sudo apt-get install build-essential tk-dev libncurses5-dev libncursesw5-dev libreadline6-dev libdb5.3-dev libgdbm-dev libsqlite3-dev libssl-dev libbz2-dev libexpat1-dev liblzma-dev zlib1g-dev libopenjp2-7 libtiff5 libffi-dev libjpeg-dev bluetooth libbluetooth-dev -y
sudo pip3 install pybluez
wget https://www.python.org/ftp/python/3.9.10/Python-3.9.10.tgz
tar xzvf Python-3.9.10.tgz
cd Python-3.9.10/
./configure --enable-optimizations
make -j 8
sudo make install
```

disabilitato l'avvio automatico e riavviato il rasp.

ho fatto la solita procedura per aggiornare ma ottengo

```
pi@domotic:~ $ cd /srv/homeassistant
pi@domotic:/srv/homeassistant $ python3.9 -m venv .
Error: [Errno 13] Permission denied: '/srv/homeassistant/lib/python3.9'
```

ed effettivamente nella cartella /srv/homeassistant/lib/ c'è solo python3.7

nonostante 

```
pi@domotic:~/Python-3.9.10 $ python3 --version
Python 3.9.10
```
L'installazione di python3.9.10 è finita così

```
WARNING: Value for scheme.headers does not match. Please report this to <https://github.com/pypa/pip/issues/10151>
distutils: /usr/local/include/python3.9/UNKNOWN
sysconfig: /home/pi/Python-3.9.10/Include/UNKNOWN
WARNING: Additional context:
user = False
home = None
root = '/'
prefix = None
Looking in links: /tmp/tmpy0zkqssh
Requirement already satisfied: setuptools in /usr/local/lib/python3.9/site-packages (58.1.0)
Requirement already satisfied: pip in /usr/local/lib/python3.9/site-packages (22.1.2)
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
pi@domotic:~/Python-3.9.10 $ 
```
e non credo sia un errore.

cosa può essere??

**Comment by sp2003** [comment]
risolta seguendo quest'altra [https://indomus.it/guide/installare-python-3-9-x-o-superiori-su-raspberry-pi-os-e-ripristinare-home-assistant-core-su-venv/)](https://indomus.it/guide/installare-python-3-9-x-o-superiori-su-raspberry-pi-os-e-ripristinare-home-assistant-core-su-venv/)
in sostanza invece di 

`pi@domotic:/srv/homeassistant $ python3.9 -m venv .`

andava fatto

`pi@domotic:/srv/homeassistant $ python3 -m venv .`

**Reply by snakuzzo** [reply]
`python3` è semplicemente un symlink. Se punta a `python3.9` è la stessa cosa. Puoi semplicemente verificare la `version` che ti restituisce...

```
bash-5.1# python3 --version
Python 3.9.12
bash-5.1# python3.9 --version
Python 3.9.12
```

...oppure puoi verificare dove ti porta il symlink

```
bash-5.1# which -a python3
/usr/local/bin/python3
bash-5.1# ls -lart /usr/local/bin/python3
lrwxrwxrwx    1 root     root             9 May 13 12:20 /usr/local/bin/python3 -> python3.9
```

Considera che il symlink cambia a seconda di come andrai a compilare la nuova versione di Python.

**Reply by snakuzzo** [reply]
Ad ogni modo io ti consiglierei di buttare via questo tipo di installazione se non sei pratico...e di utilizzare Docker.

**Reply by sp2003** [reply]
> Ad ogni modo io ti consiglierei di buttare via questo tipo di installazione se non sei pratico...e di utilizzare Docker.

ci sto pensando seriamente, ma mi piace perdere la testa e provare a risolvere. Al max chiedo a google, ai forum o a Diego :P

**Reply by snakuzzo** [reply]
Sì ok :D, ma visto quello che hai fatto prima, hai risolto senza capirne il motivo

**Reply by sp2003** [reply]
ma infatti non capivo il motivo del problema.
Cioè, dopo aver installato Python3.9.10, credevo di trovare in /srv/homeassistant/lib la nuova installazione, invece c'era la vecchia 3.7
Ovviamente si è aggiornata dopo che ho ricreato il nuovo ambiente con il binario aggiornato.

cmq ora mi sono accorto di un nuovo problema.
La energy dashboard non c'è e come integrazione non compare tra le opzioni.....
